File: SplitFree/Core/Sync/Supabase.swift
import Foundation
import Security

/// Where the backend lives, and whether there is one at all.
///
/// Both values come from the app's Info.plist so the open-source repo doesn't
/// carry anyone's project keys. When they're absent, `isConfigured` is false and
/// the whole sync feature stays switched off: the app is exactly the local-only
/// app it was before, which is the point of making accounts optional.
enum SupabaseConfig {
    /// The project's base URL.
    ///
    /// The scheme is optional in the config file and added here when missing.
    /// That is not politeness: `//` starts a comment in an xcconfig, so a pasted
    /// `https://…` is silently truncated to `https:` and every request fails
    /// with nothing on screen to explain it. Accepting a bare host removes the
    /// trap rather than documenting it.
    static let url: URL? = {
        guard var raw = Bundle.main.object(forInfoDictionaryKey: "SupabaseURL") as? String else {
            return nil
        }
        raw = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        while raw.hasSuffix("/") { raw.removeLast() }
        guard !raw.isEmpty, raw != "https:", raw != "http:" else { return nil }
        if !raw.lowercased().hasPrefix("http") {
            // Loopback means a stack running on this machine, which serves plain
            // HTTP. Inferring it is what lets the scheme stay out of the config
            // file entirely, which is the point: `//` starts a comment there.
            let isLoopback = raw.hasPrefix("127.0.0.1")
                || raw.hasPrefix("localhost")
                || raw.hasPrefix("[::1]")
            raw = (isLoopback ? "http://" : "https://") + raw
        }
        return URL(string: raw)
    }()

    static let anonKey: String? = {
        guard let key = Bundle.main.object(forInfoDictionaryKey: "SupabaseAnonKey") as? String
        else { return nil }
        let trimmed = key.trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? nil : trimmed
    }()

    static var isConfigured: Bool { url != nil && anonKey != nil }

}

// MARK: - Session

struct SupabaseSession: Codable, Equatable {
    var accessToken: String
    var refreshToken: String
    var expiresAt: Date
    var userID: String

    /// Treated as expired a minute early, so a request never leaves with a token
    /// that dies in flight.
    var isExpired: Bool { Date() >= expiresAt.addingTimeInterval(-60) }
}

/// Session storage. The Keychain rather than UserDefaults, because a refresh
/// token is a long-lived credential.
enum SessionStore {
    private static let service = "com.splitfree.session"
    private static let account = "supabase"

    static func save(_ session: SupabaseSession) {
        guard let data = try? JSONEncoder().encode(session) else { return }
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account,
        ]
        SecItemDelete(query as CFDictionary)
        var attributes = query
        attributes[kSecValueData as String] = data
        attributes[kSecAttrAccessible as String] = kSecAttrAccessibleAfterFirstUnlock
        SecItemAdd(attributes as CFDictionary, nil)
    }

    static func load() -> SupabaseSession? {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account,
            kSecReturnData as String: true,
            kSecMatchLimit as String: kSecMatchLimitOne,
        ]
        var item: CFTypeRef?
        guard SecItemCopyMatching(query as CFDictionary, &item) == errSecSuccess,
              let data = item as? Data
        else { return nil }
        return try? JSONDecoder().decode(SupabaseSession.self, from: data)
    }

    static func clear() {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account,
        ]
        SecItemDelete(query as CFDictionary)
    }
}

// MARK: - Errors

enum SupabaseError: LocalizedError {
    case notConfigured
    case notSignedIn
    case http(status: Int, message: String)
    case decoding(String)
    case network(String)

    var errorDescription: String? {
        switch self {
        case .notConfigured:
            String(localized: "This build isn't set up for syncing.")
        case .notSignedIn:
            String(localized: "Sign in to sync.")
        case .http(let status, let message):
            message.isEmpty
                ? String(format: String(localized: "The server returned an error (%lld).", comment: "HTTP status"), status)
                : message
        case .decoding:
            String(localized: "The server sent something SplitFree didn't understand.")
        case .network(let message):
            message
        }
    }
}

// MARK: - Client

/// A small, dependency-free Supabase client: GoTrue for auth, PostgREST for
/// data, and RPC for the sync and invite functions.
///
/// Only the handful of endpoints this app actually uses are implemented. That's
/// a few hundred lines against pulling in an SDK, and it keeps the promise in
/// the README that the app has no third-party dependencies.
actor SupabaseClient {
    static let shared = SupabaseClient()

    private let session: URLSession
    private var current: SupabaseSession?
    /// Serialises refreshes so ten parallel requests don't each spend the
    /// refresh token, which would invalidate the other nine.
    private var refreshTask: Task<SupabaseSession, Error>?

    init(session: URLSession = .shared) {
        self.session = session
        self.current = SessionStore.load()
    }

    var isSignedIn: Bool { current != nil }
    var userID: String? { current?.userID }

    func restoreSession() -> SupabaseSession? {
        current = SessionStore.load()
        return current
    }

    // MARK: Identity

    /// Gets an identity without asking anybody for anything.
    ///
    /// Sharing needs row level security to have something to key on, and that
    /// is all it needs. There is no email, no password and no screen: the first
    /// time somebody shares or joins, the device quietly becomes a user.
    ///
    /// The trade is recovery. Lose the device and the session goes with it, and
    /// a friend has to send a fresh code. On iPhone the Keychain usually
    /// survives a restore, so in practice this bites rarely.
    @discardableResult
    func signInAnonymously() async throws -> SupabaseSession {
        if let current, !current.isExpired { return current }
        let data = try await request(path: "/auth/v1/signup", method: "POST", body: [:], authorized: false)
        return try store(tokenResponse: data)
    }

    func signOut() {
        current = nil
        refreshTask = nil
        SessionStore.clear()
    }

    private func store(tokenResponse data: Data) throws -> SupabaseSession {
        struct Payload: Decodable {
            let access_token: String
            let refresh_token: String
            let expires_in: Double?
            let user: User?
            struct User: Decodable { let id: String }
        }
        guard let payload = try? JSONDecoder().decode(Payload.self, from: data) else {
            throw SupabaseError.decoding("token")
        }
        let session = SupabaseSession(
            accessToken: payload.access_token,
            refreshToken: payload.refresh_token,
            expiresAt: Date().addingTimeInterval(payload.expires_in ?? 3600),
            userID: payload.user?.id ?? ""
        )
        current = session
        SessionStore.save(session)
        return session
    }

    private func validSession() async throws -> SupabaseSession {
        guard let existing = current else { return try await signInAnonymously() }
        guard existing.isExpired else { return existing }

        if let refreshTask { return try await refreshTask.value }
        let task = Task<SupabaseSession, Error> {
            let data = try await request(
                path: "/auth/v1/token?grant_type=refresh_token",
                method: "POST",
                body: ["refresh_token": existing.refreshToken],
                authorized: false
            )
            return try store(tokenResponse: data)
        }
        refreshTask = task
        defer { refreshTask = nil }
        do {
            return try await task.value
        } catch {
            // A dead refresh token is not worth surfacing when the identity is
            // disposable. Take a fresh one.
            signOut()
            return try await signInAnonymously()
        }
    }

    // MARK: Data

    /// Calls a Postgres function. Everything the sync engine does goes through
    /// one of these.
    func rpc(_ name: String, arguments: [String: Any] = [:]) async throws -> Data {
        try await request(path: "/rest/v1/rpc/\(name)", method: "POST", body: arguments, authorized: true)
    }

    /// Inserts or updates rows, keyed by primary key.
    func upsert(table: String, rows: [[String: Any]], onConflict: String? = nil) async throws {
        guard !rows.isEmpty else { return }
        var path = "/rest/v1/\(table)"
        if let onConflict { path += "?on_conflict=\(onConflict)" }
        _ = try await request(
            path: path,
            method: "POST",
            bodyArray: rows,
            authorized: true,
            extraHeaders: ["Prefer": "resolution=merge-duplicates,return=minimal"]
        )
    }

    // MARK: Transport

    private func request(
        path: String,
        method: String,
        body: [String: Any]? = nil,
        bodyArray: [[String: Any]]? = nil,
        authorized: Bool,
        extraHeaders: [String: String] = [:]
    ) async throws -> Data {
        guard let base = SupabaseConfig.url, let anonKey = SupabaseConfig.anonKey else {
            throw SupabaseError.notConfigured
        }

        var urlRequest = URLRequest(url: base.appendingPathComponent(path))
        // `appendingPathComponent` escapes the query string, so rebuild when one
        // is present.
        if path.contains("?"), let direct = URL(string: base.absoluteString + path) {
            urlRequest = URLRequest(url: direct)
        }
        urlRequest.httpMethod = method
        urlRequest.timeoutInterval = 30
        urlRequest.setValue(anonKey, forHTTPHeaderField: "apikey")
        urlRequest.setValue("application/json", forHTTPHeaderField: "Content-Type")
        for (key, value) in extraHeaders {
            urlRequest.setValue(value, forHTTPHeaderField: key)
        }

        // Authorization carries a user's session and nothing else. The older
        // anon key was a JWT and could stand in for one, which is why so much
        // sample code sends it here; the newer publishable keys are not JWTs, and
        // a server asked to parse one as a token rejects the request. The apikey
        // header is what identifies the project, and it is set above either way.
        if authorized {
            let session = try await validSession()
            urlRequest.setValue("Bearer \(session.accessToken)", forHTTPHeaderField: "Authorization")
        }

        if let body {
            urlRequest.httpBody = try? JSONSerialization.data(withJSONObject: body)
        } else if let bodyArray {
            urlRequest.httpBody = try? JSONSerialization.data(withJSONObject: bodyArray)
        }

        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await session.data(for: urlRequest)
        } catch {
            throw SupabaseError.network(error.localizedDescription)
        }

        guard let http = response as? HTTPURLResponse else {
            throw SupabaseError.network(String(localized: "No response from the server."))
        }
        guard (200..<300).contains(http.statusCode) else {
            throw SupabaseError.http(status: http.statusCode, message: Self.errorMessage(from: data))
        }
        return data
    }

    /// Supabase reports failures in a few different shapes depending on which
    /// component answered.
    private static func errorMessage(from data: Data) -> String {
        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else { return "" }
        for key in ["msg", "message", "error_description", "error", "hint", "details"] {
            if let value = json[key] as? String, !value.isEmpty { return value }
        }
        return ""
    }
}

File: android/app/src/main/kotlin/com/splitfree/sync/Supabase.kt
package com.splitfree.sync

import android.content.Context
import android.content.SharedPreferences
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey
import com.splitfree.BuildConfig
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import java.io.BufferedReader
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLEncoder

/**
 * Where the backend lives, and whether there is one at all.
 *
 * Both values come from build config so the open-source repo carries nobody's
 * project keys. When they are blank, [isConfigured] is false and the whole sync
 * feature stays off: the app is exactly the local-only app it was before, which
 * is the point of making accounts optional.
 */
object SupabaseConfig {
    /**
     * The project's base URL. The scheme is optional in the config file and
     * added here when missing, so the same value works in either platform's
     * config format.
     */
    val url: String
        get() {
            val raw = BuildConfig.SUPABASE_URL.trim().trimEnd('/')
            if (raw.isBlank() || raw == "https:" || raw == "http:") return ""
            if (raw.startsWith("http", ignoreCase = true)) return raw
            // Loopback means a stack running on this machine, which serves plain
            // HTTP. On Android the emulator reaches the host at 10.0.2.2.
            val isLoopback = raw.startsWith("127.0.0.1") ||
                raw.startsWith("localhost") ||
                raw.startsWith("10.0.2.2") ||
                raw.startsWith("[::1]")
            return (if (isLoopback) "http://" else "https://") + raw
        }

    val anonKey: String get() = BuildConfig.SUPABASE_ANON_KEY.trim()

    val isConfigured: Boolean get() = url.isNotBlank() && anonKey.isNotBlank()

}

class SupabaseException(message: String) : Exception(message)

data class Session(
    val accessToken: String,
    val refreshToken: String,
    val expiresAtMillis: Long,
    val userId: String,
) {
    /** Treated as expired a minute early, so a request never leaves with a token that dies in flight. */
    val isExpired: Boolean get() = System.currentTimeMillis() >= expiresAtMillis - 60_000
}

/**
 * A small, dependency-free Supabase client: GoTrue for auth, PostgREST for data,
 * and RPC for the sync and invite functions.
 *
 * Only the endpoints this app actually uses are implemented. That is a few
 * hundred lines against pulling in an SDK, and it keeps the dependency list
 * short enough to read.
 */
class SupabaseClient(context: Context) {

    private val appContext = context.applicationContext
    private val refreshLock = Mutex()

    /**
     * A refresh token is a long-lived credential, so it goes in encrypted
     * preferences rather than ordinary ones.
     */
    private val prefs: SharedPreferences by lazy {
        runCatching {
            EncryptedSharedPreferences.create(
                appContext,
                "splitfree.session",
                MasterKey.Builder(appContext).setKeyScheme(MasterKey.KeyScheme.AES256_GCM).build(),
                EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
                EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM,
            )
        }.getOrElse {
            // Some devices have a broken keystore. Signing out is better than
            // refusing to launch.
            appContext.getSharedPreferences("splitfree.session", Context.MODE_PRIVATE)
        }
    }

    @Volatile private var current: Session? = null

    init {
        current = readSession()
    }

    val isSignedIn: Boolean get() = current != null
    val userId: String? get() = current?.userId

    // MARK: - Identity

    /**
     * Gets an identity without asking anybody for anything.
     *
     * Sharing needs row level security to have something to key on, and that is
     * all it needs. No email, no password, no screen.
     */
    suspend fun signInAnonymously(): Session {
        current?.let { if (!it.isExpired) return it }
        return storeSession(JSONObject(request("/auth/v1/signup", "POST", JSONObject(), authorized = false)))
    }

    fun signOut() {
        current = null
        prefs.edit().clear().apply()
    }

    private fun storeSession(payload: JSONObject): Session {
        val access = payload.optString("access_token")
        val refresh = payload.optString("refresh_token")
        if (access.isBlank() || refresh.isBlank()) throw SupabaseException("The server sent an unusable session.")
        val session = Session(
            accessToken = access,
            refreshToken = refresh,
            expiresAtMillis = System.currentTimeMillis() + payload.optLong("expires_in", 3600) * 1000,
            userId = payload.optJSONObject("user")?.optString("id").orEmpty(),
        )
        writeSession(session)
        return session
    }

    private fun writeSession(session: Session) {
        current = session
        prefs.edit()
            .putString("access", session.accessToken)
            .putString("refresh", session.refreshToken)
            .putLong("expires", session.expiresAtMillis)
            .putString("user", session.userId)
            .apply()
    }

    private fun readSession(): Session? {
        val access = prefs.getString("access", null) ?: return null
        val refresh = prefs.getString("refresh", null) ?: return null
        return Session(
            accessToken = access,
            refreshToken = refresh,
            expiresAtMillis = prefs.getLong("expires", 0),
            userId = prefs.getString("user", "").orEmpty(),
        )
    }

    private suspend fun validSession(): Session {
        val existing = current ?: return signInAnonymously()
        if (!existing.isExpired) return existing

        return refreshLock.withLock {
            val latest = current ?: return@withLock signInAnonymously()
            if (!latest.isExpired) return@withLock latest
            try {
                storeSession(
                    JSONObject(
                        request(
                            "/auth/v1/token?grant_type=refresh_token",
                            "POST",
                            JSONObject().put("refresh_token", latest.refreshToken),
                            authorized = false,
                        )
                    )
                )
            } catch (error: Exception) {
                // A dead refresh token is not worth surfacing when the identity
                // is disposable. Take a fresh one.
                signOut()
                signInAnonymously()
            }
        }
    }

    // MARK: - Data

    suspend fun rpc(name: String, arguments: JSONObject = JSONObject()): String =
        request("/rest/v1/rpc/$name", "POST", arguments, authorized = true)

    suspend fun upsert(table: String, rows: JSONArray, onConflict: String) {
        if (rows.length() == 0) return
        request(
            path = "/rest/v1/$table?on_conflict=$onConflict",
            method = "POST",
            body = rows,
            authorized = true,
            extraHeaders = mapOf("Prefer" to "resolution=merge-duplicates,return=minimal"),
        )
    }

    // MARK: - Transport

    private suspend fun request(
        path: String,
        method: String,
        body: Any? = null,
        authorized: Boolean,
        extraHeaders: Map<String, String> = emptyMap(),
    ): String {
        if (!SupabaseConfig.isConfigured) throw SupabaseException("This build isn't set up for syncing.")

        // Authorization carries a user's session and nothing else. The older
        // anon key was a JWT and could stand in for one; the newer publishable
        // keys are not JWTs, and a server asked to parse one as a token rejects
        // the request. The apikey header identifies the project either way.
        val token = if (authorized) validSession().accessToken else null

        return withContext(Dispatchers.IO) {
            val connection = (URL(SupabaseConfig.url + path).openConnection() as HttpURLConnection).apply {
                requestMethod = method
                connectTimeout = 15_000
                readTimeout = 30_000
                setRequestProperty("apikey", SupabaseConfig.anonKey)
                if (token != null) setRequestProperty("Authorization", "Bearer $token")
                setRequestProperty("Content-Type", "application/json")
                extraHeaders.forEach { (key, value) -> setRequestProperty(key, value) }
                if (body != null) {
                    doOutput = true
                    outputStream.use { it.write(body.toString().toByteArray(Charsets.UTF_8)) }
                }
            }

            try {
                val code = connection.responseCode
                val stream = if (code in 200..299) connection.inputStream else connection.errorStream
                val text = stream?.bufferedReader()?.use(BufferedReader::readText).orEmpty()
                if (code !in 200..299) throw SupabaseException(errorMessage(text, code))
                text
            } finally {
                connection.disconnect()
            }
        }
    }

    /** Supabase reports failures in a few shapes, depending on which component answered. */
    private fun errorMessage(body: String, code: Int): String {
        val fallback = "The server returned an error ($code)."
        return runCatching {
            val json = JSONObject(body)
            listOf("msg", "message", "error_description", "error", "hint", "details")
                .firstNotNullOfOrNull { json.optString(it).takeIf { value -> value.isNotBlank() } }
                ?: fallback
        }.getOrDefault(fallback)
    }

    companion object {
        const val SIGNED_OUT = "Sign in to sync."
    }
}
